### PR TITLE
(prometheus) dedupe same timestamp annotation

### LINKED ISSUE #15064: "Series value as timestamp" annotations should be deduplicated
### What Grafana version are you using?
5.4.2
### What datasource are you using?
Prometheus
### What OS are you running grafana on?
Linux
### What did you do?
Created a Prometheus query-based annotation that had the "Series value as timestamp" option checked (see #11789). The annotation was supposed to display a marker for every instance restart.
### What was the expected result?
I was expecting to see a couple of markers for the 2 restarts that happened during the past 24 hours, after a second or two of loading.
### What happened instead?
Visually I got the expected result, but loading took minutes rather than seconds and AFAICT there was one marker per sample returned by Prometheus (all drawn on top of each other, in the 2 locations) rather than only 2 markers. This caused the Chrome tab to peg the CPU at 100% for minutes and the tab was unresponsive for the duration.

Ideally, for annotations with "Series value as timestamp" checked, the values would first be de-duplicated and only then turned into markers on the graph. There is no good reason to draw 1k markers with the exact same title, tags and text, on top of each other.

### PR DESCRIPTION
**Pull request by mtanda**
Fix https://github.com/grafana/grafana/issues/15064

### CODE REVIEW

File: public/app/plugins/datasource/prometheus/datasource.ts [hunk 1]
```diff
@@ -365,6 +366,10 @@ export class PrometheusDatasource implements DataSourceApi<PromQuery> {
             };
 
             if (annotation.useValueForTime) {
+              if (dupCheck[value[1]]) {
+                continue;
+              }
+              dupCheck[value[1]] = true;
```
Comment (torkelo) [inline public/app/plugins/datasource/prometheus/datasource.ts]: looks like value[1] is used in a at least 3 places in this loop, maybe extract value[1] and to variables?

Comment (mtanda) [reply public/app/plugins/datasource/prometheus/datasource.ts]: Of course, I did!

File: public/app/plugins/datasource/prometheus/datasource.ts [hunk 2]
```diff
@@ -365,7 +366,12 @@ export class PrometheusDatasource implements DataSourceApi<PromQuery> {
             };
 
             if (annotation.useValueForTime) {
-              event['time'] = Math.floor(parseFloat(value[1]));
+              const timestampValue = Math.floor(parseFloat(value[1]));
+              if (dupCheck[timestampValue]) {
+                continue;
+              }
+              dupCheck[timestampValue] = true;
+              event['time'] = timestampValue;
```
Comment (torkelo) [inline public/app/plugins/datasource/prometheus/datasource.ts]: why index lookup? shoulb be event.time = eventTime

Comment (mtanda) [reply public/app/plugins/datasource/prometheus/datasource.ts]: I don't remember why I did, anyway It should be `event.time` now.

File: public/app/plugins/datasource/prometheus/datasource.ts [hunk 3]
```diff
@@ -354,20 +354,26 @@ export class PrometheusDatasource implements DataSourceApi<PromQuery> {
           })
           .value();
 
+        const dupCheck = {};
         for (const value of series.values) {
           const valueIsTrue = value[1] === '1'; // e.g. ALERTS
           if (valueIsTrue || annotation.useValueForTime) {
-            const event = {
+            const event: any = {
```
Comment (torkelo) [inline public/app/plugins/datasource/prometheus/datasource.ts]: seems unnecessary to define it as any & loose all typing, think we need an annotation event type, and add type definition for the annotationQuery method, like we have for query method. If you want me to do that I can try find time the next few days.

Comment (mtanda) [reply public/app/plugins/datasource/prometheus/datasource.ts]: I think so too, we need annotation event type. If you add it, really helpful for me.

**Pull request merged**